### PR TITLE
53021 inject user data in fsr jobs

### PR DESCRIPTION
**Pull request by kjsuarez**
## Summary

The FSR service takes a user as an input. This can become an issue when long running jobs try to trigger the service because the user object may become stale before the job can run. This team does not control the ttl of the user model and so we've opted to pass the necessary user data into the job as parameters rather than hoping that the user uuid we pass in is still valid.

## Testing done

Existing specs are green and additional specs have been written to confirm that the added parameters are being utilized.

## What areas of the site does it impact?
- The financial status report

## Acceptance criteria

- [ ]  Effected unit test pass
- [ ]  No error nor warning in the console.
- [ ]  The FSR service can submit vbs and vhs FSRs with or without the additional user parameters

### CODE REVIEW

**Review by ScottyRJames — DISMISSED** [review]
👍

**Review by ryan-mcneil — APPROVED** [review]
LGTM

**Comment by bosawt** [comment]
I'm out of office this week, but wanted to comment that we really shouldn't be sending ssn as a sidekiq param. Please hold onto this until next week and I'll help you with a better way that'll still accomplish what you want.

**Comment by joeniquette** [comment]
@kjsuarez Can you use ICN instead of SSN here?

File: app/models/user.rb [hunk 1]
```diff
@@ -421,6 +421,16 @@ def relationships
     @relationships ||= get_relationships_array
   end
 
+  def identity_serial
+    {
+      'uuid' => uuid,
+      'first_name' => first_name,
+      'last_name' => last_name,
+      'email' => email,
+      'account_id' => account.id
```
Comment (bosawt) [inline app/models/user.rb:430]: Please don't use `Account` model for new functionality, we're deprecating this. `user.user_account.id` is the appropriate alternative to use (and you can grab the ICN off of that).

File: app/models/account.rb [hunk 2]
```diff
@@ -64,6 +64,14 @@ def self.lookup_by_user_account_uuid(user_uuid)
     Account.where(icn: user_account.icn).first
   end
 
+  def mpi_profile
+    resp = MPI::Service.new.find_profile_by_identifier(
```
Comment (bosawt) [inline app/models/account.rb:68]: What would you estimate the magnitude of the number of jobs with this being? Calling MPI is a nice solution to not have to provide SSN across sidekiq params, but want to make sure we aren't DDOS'ing the service

File: app/models/user.rb [hunk 3]
```diff
@@ -421,6 +421,16 @@ def relationships
     @relationships ||= get_relationships_array
   end
 
+  def identity_serial
+    {
+      'uuid' => uuid,
```
Comment (bosawt) [inline app/models/user.rb:426]: If we call MPI to get SSN, might as well call it to get the rest of the PII data as well

**Review by bosawt — CHANGES_REQUESTED** [review]
Made some comments that are hopefully helpful. I wanted to describe two possibilities for a solution here:

- One solution is the one you're moving towards here, which is to pass the ICN as a sidekiq param, and then calling MPI to get data that you need. This solution is totally fine (you'll be able to replay the jobs, and won't have to worry about the redis going away etc), but the one thing to consider is the number of jobs being enqueued, as we definitely don't want to bring the MPI service down with a massive number of requests (just let me know, this may or may not be a problem depending on usage)

- A second solution is to create your own redis just for this job, which represents the serialized user data. As long as this redis entry is temporary, you can store the SSN, and other PII in it and not have to worry about compliance issues. The nice thing with this solution is that you can store exactly what you want and know that you will be pulling the same data on the sidekiq side (won't have to worry about hitting MPI too hard or anything). You would then pass the key to the redis entry across sidekiq, then retrieve it (and delete the redis entry) when the job was run and passed successfully

File: app/models/account.rb (hunk 2 above)
Comment (kjsuarez) [reply app/models/account.rb:68]: ≈ 100/day

File: lib/debt_management_center/financial_status_report_service.rb [hunk 4]
```diff
@@ -256,21 +256,22 @@ def set_certification_date(form)
       form['applicantCertifications']['veteranDateSigned'] = date_formatted if form['applicantCertifications']
     end
 
-    def send_confirmation_email(template_id)
+    def send_confirmation_email(template_id, user_params = {})
       return unless Flipper.enabled?(:fsr_confirmation_email)
 
-      email = @user.email&.downcase
+      email = @user&.email&.downcase || user_params['email']
```
[the remark this answers is not in the document]
Comment (kjsuarez) [reply lib/debt_management_center/financial_status_report_service.rb:262]: Using user params only as a fall back would solve your concern about spamming MPI.

**Comment by kjsuarez** [comment]
> * A second solution is to create your own redis just for this job, which represents the serialized user data. As long as this redis entry is temporary, you can store the SSN, and other PII in it and not have to worry about compliance issues. The nice thing with this solution is that you can store exactly what you want and know that you will be pulling the same data on the sidekiq side (won't have to worry about hitting MPI too hard or anything). You would then pass the key to the redis entry across sidekiq, then retrieve it (and delete the redis entry) when the job was run and passed successfully

If I'm able to store PII in a redis store than I think that would definitely be the cleaner option in the long run. I'm going to make a separate branch for that  effort and get back to you.

**Comment by kjsuarez** [comment]
Work implemented in [this PR](https://github.com/department-of-veterans-affairs/vets-api/pull/12928/commits/22795a78d2df97d05df3b42794ef5777a0ca5d21)

**Pull request closed without merging**